pycbc/noise/reproduceable.py

```diff
@@ -0,0 +1,142 @@
+# Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# =============================================================================
+#
+#                                   Preamble
+#
+# =============================================================================
+#
+import numpy, pycbc.psd
+from pycbc.types import TimeSeries
+from numpy.random import RandomState
+
+# These need to be constant to be able to recover identical results.
+# The hope is that nobody needs a higher resolution
+SAMPLE_RATE = 16384
+BLOCK_SIZE = 100
+FILTER_LENGTH = 128
+
+def block(seed):
+    """ Return block of normal random numbers
+    
+    Parameters
+    ----------
+    seed : {None, int}
+        The seed to generate the noise.
+
+    Returns
+    --------
+    noise : numpy.ndarray
+        Array of random numbers
+    """ 
+    num = SAMPLE_RATE * BLOCK_SIZE
+    rng = RandomState(seed % 2**32)
+    variance = SAMPLE_RATE / 2
+    return rng.normal(size=num, scale=variance**0.5)
+
+def normal(start, end, seed=0):
+    """ Generate data with a white Gaussian (normal) distribution
+
+    Parameters
+    ----------
+    start_time : int
+        Start time in GPS seconds to generate noise
+    end_time : int
+        End time in GPS seconds to generate nosie
+    seed : {None, int}
+        The seed to generate the noise.
+
+    Returns
+    --------
+    noise : TimeSeries
+        A TimeSeries containing gaussian noise
+    """
+    # This is reproduceable because we used fixed seeds from known values
+    s = int(start / BLOCK_SIZE)
+    e = int(end / BLOCK_SIZE)
+
+    sv = RandomState(seed).randint(-2**50, 2**50)
+    data = numpy.concatenate([block(i + sv) for i in numpy.arange(s, e + 1, 1)])
+    return TimeSeries(data, delta_t=1.0 / SAMPLE_RATE, epoch=start)
+
+def colored_noise(psd, start_time, end_time, seed=0, low_frequency_cutoff=1.0):
+    """ Create noise from a PSD
+
+    Return noise from the chosen PSD. Note that if unique noise is desired
+    a unique seed should be provided.
+
+    Parameters
+    ----------
+    psd : pycbc.types.FrequencySeries
+        PSD to color the noise
+    start_time : int
+        Start time in GPS seconds to generate noise
+    end_time : int
+        End time in GPS seconds to generate nosie
+    seed : {None, int}
+        The seed to generate the noise.
+    low_frequency_cutof : {10.0, float}
+        The low frequency cutoff to pass to the PSD generation.
+
+    Returns
+    --------
+    noise : TimeSeries
+        A TimeSeries containing gaussian noise colored by the given psd.
+    """ 
+    white_noise = normal(start_time - FILTER_LENGTH, end_time + FILTER_LENGTH,
+                          seed=seed)   
+    flen = (SAMPLE_RATE / psd.delta_f) / 2 + 1
+    psd.resize(flen)
```

Comment:
Yeah, maybe. It just seems wasteful to duplicate memory here ... But probably it doesn't matter overall.